cmd/controller-manager/app/config/types.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/restmapper"
+	ccmconfig "k8s.io/kubernetes/cmd/cloud-controller-manager/app/apis/config"
+	"k8s.io/kubernetes/pkg/controller"
+
+	"time"
+)
+
+// ControllerLoopMode is the kube-controller-manager's mode of running controller loops that are cloud provider dependent
+type ControllerLoopMode int
+
+const (
+	// IncludeCloudLoops means the kube-controller-manager include the controller loops that are cloud provider dependent
+	IncludeCloudLoops ControllerLoopMode = iota
+	// ExternalLoops means the kube-controller-manager exclude the controller loops that are cloud provider dependent
+	ExternalLoops
+)
+
+// ControllerContext defines the context object for controller
+type ControllerContext struct {
+	// ClientBuilder will provide a client for this controller to use
+	ClientBuilder controller.ControllerClientBuilder
+
+	// InformerFactory gives access to informers for the controller.
+	InformerFactory informers.SharedInformerFactory
+
+	// ObjectOrMetadataInformerFactory gives access to informers for typed resources
+	// and dynamic resources by their metadata. All generic controllers currently use
+	// object metadata - if a future controller needs access to the full object this
+	// would become GenericInformerFactory and take a dynamic client.
+	ObjectOrMetadataInformerFactory controller.InformerFactory
+
+	ComponentConfig ccmconfig.CloudControllerManagerConfiguration
```

Comment:
Can we get rid of this configuration? If we want to make this a generic controller manager ControllerContext then is shouldn't have anything CCM specific in it.

Reply:
Thanks for pointing out. Updated